torch/csrc/autograd/record_function.cpp

```diff
@@ -26,39 +27,86 @@ void popCallback() {
   }
   start_callbacks.pop_back();
   end_callbacks.pop_back();
+  needs_inputs_flags.pop_back();
   has_callbacks = !start_callbacks.empty();
+  cb_needs_inputs = std::any_of(
+    needs_inputs_flags.begin(), needs_inputs_flags.end(),
+    [](bool v) { return v; });
 }
 
-RecordFunction::RecordFunction(Function* fn, GetPackedInputsCallback cb) {
+TORCH_API RecordFunctionState recordFunctionState() {
+  if (!has_callbacks) {
+    return RecordFunctionState::NO_CALLBACKS;
+  }
+  if (cb_needs_inputs) {
+    return RecordFunctionState::NAME_AND_INPUTS;
+  }
+  return RecordFunctionState::NAME_ONLY;
+}
+
+RecordFunction::RecordFunction(Function* fn) {
+  before(fn);
+}
+
+RecordFunction::RecordFunction(
+    std::string name, int64_t sequence_nr) {
+  before(name, c10::nullopt, sequence_nr);
+}
+
+RecordFunction::RecordFunction(
+    const char* name, int64_t sequence_nr) {
+  before(name, c10::nullopt, sequence_nr);
+}
+
+void RecordFunction::before(
+    Function* fn, c10::optional<at::ArrayRef<c10::IValue>> inputs) {
   if (!has_callbacks) {
     return;
   }
+  if (initialized_) {
+    throw std::runtime_error("Double initializing RecordFunction");
+  }
   fn_ = fn;
   name_ = StringView(fn->name());
   sequence_nr_ = fn->sequence_nr();
-  inputs_cb_ = cb;
+  if (inputs) {
+    copyInputs(*inputs);
+  }
+  initialized_ = true;
   processCallbacks();
 }
 
-RecordFunction::RecordFunction(
-    std::string name, int64_t sequence_nr, GetPackedInputsCallback cb) {
+void RecordFunction::before(
+    std::string name, c10::optional<at::ArrayRef<c10::IValue>> inputs, int64_t sequence_nr) {
   if (!has_callbacks) {
     return;
   }
+  if (initialized_) {
+    throw std::runtime_error("Double initializing RecordFunction");
+  }
   name_ = StringView(std::move(name));
```

Comment:
not for this diff - but doesn't it capture a string view to a temporary?

Reply:
StringView owns the std::string, or did you mean smth else?

Reply:
oh, nevermind, I didn't read :)